text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,284 @@
+ 
+FPC is a framework to develop and run chaincodes with strong privacy guarantees.
+With FPC, a chaincode can process transaction arguments and state without revealing/exposing the contents to anybody, including the endorsing peers. That is, clients can encrypt the content of a transaction in a way only the FPC chaincode can decrypt and then process it. The resulting state updates are encrypted by default by the FPC chaincode with keys only known to the chaincode. 
+
+As a result, other parties (both clients and peers) cannot examine the state of the transactions either in the chaincode or on the ledger. Hence, FPC helps to protects sensitive data even from compromised endorsing peers and other unauthorised parties.
+FPC provides the capabilities to verify that a chaincode and the data are protected, that is, Fabric peers and clients can receive cryptographic assurance that the correct chaincode is being run without tampering inside the TEE by means of a hardware-based remote attestation. 
+
+This new model of trust for Fabric smart contracts makes possible high-stakes markets such as private voting systems and sealed-bid auctions; which aren’t supported by the existing model of privacy in Fabric because of the requirement for endorsing peers to be fully trusted for confidentiality.
+ 
+A prototype implementation of FPC is available as Hyperledger Lab on github (https://github.com/hyperledger-labs/fabric-private-chaincode).
+
+# Motivation
+[motivation]: #motivation
+ 
+FPC is motivated by the many use cases in which it is desirable to embody an application in a Blockchain architecture, but where in addition to the existing Integrity assurances, the application also requires privacy. This may include private voting, sealed bid auctions, operations on sensitive data such as regulated medical or genomic data, and supply chain operations requiring contract secrecy. With Fabric's current privacy mechanisms, these use cases are not possible as they still require the endorsement nodes to be fully trusted.
+For example, the concept of channels and Private Data allows to restrict chaincode data sharing only within a group of authorized participants, still when the chaincode processes the data it is exposed to the endorsing peer in clear. In the example of a voting system, where a government may run an endorsing peer it is clear that this is not ideal. 
+ 
+A second motivation for FPC is its integrity model based on hardware-enabled remote cryptographic attestation; this model can provide similarly high confidence of integrity to the standard Fabric model of integrity through redundancy, but using less computation and communication. With TEE-based endorsement and remote attestation, a new set of endorsement policies are made possible, which can reduce the number of required endorsements and still provide sufficient assurance of integrity for many workloads.
```

Comment:
TEE-based endorsement is indeed not a complete replacement of all fabric endorsements/agreements. Any change of meta-data such as code upgrade, keys required to bootstrap security and alike will still rely on normal fabric policies for approving the corresponding meta-data state.    For endorsements of actual chaincode transactions, enclave endorsements (attestations) are sufficient, though.  Note that the n-programming model which fabric allows now with v2 works well for integrity guarantees but does not (practically) work for confidentiality guarantees. In the latter case you require that a chaincode only ever releases information according the specification at _endorsement time_ , there is no way to prevent wrong execution at _validation time_ (as for integrity) as the damage (leakage of sensitive information) already has happened before and a smart attacker would never send such leaky endorsement for validation ..

Regarding key-based endorsement: this is not a currently supported feature in FPC.

Reply:
Agreed. A peer's endorsement can be replaced with TEE-based endorsement (attestation). However,  my comment was related to the following points which suggested replacing existing endorsement policy with single/fewer TEE endorsement.
> this model can provide similarly high confidence of integrity to the standard Fabric model of integrity through redundancy, but using less computation and communication. With TEE-based endorsement and remote attestation, a new set of endorsement policies are made possible, which can reduce the number of required endorsements and still provide sufficient assurance of integrity for many workloads.

Hence, I mentioned that endorsements are not only to ensure integrity, it is also to ensure
- deterministic execution of smart-contract (as we support general-purpose language unlike Ethereum which support deterministic solidity).
- agreements (a part of consensus).
- to allow each organization to run a customized logic.

Reply:
The first point is automatically guaranteed for a particular chaincode, everybody will run the same enclave code. As mentioned, to ensure confidentiality this is almost unavoidable; in particular it practically rules out organization-private custom logic (you would have to proof that such code still guarantees the overall privacy guarantees, you would have to do that in zero-knowledge if you want to keep logic secret and this would have to hold for potentially arbitrary code. Theoretically of course possible but not very practical).    For more restricted classes of customization the problem could be simpler: What concretely do you have in mind when you say customized logic?